src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointJsonPreparationEmitter.cs

```diff
@@ -0,0 +1,57 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Http.RequestDelegateGenerator.StaticRouteHandlerModel.Emitters;
+
+internal static class EndpointJsonPreparationEmitter
+{
+    internal static void EmitJsonPreparation(this Endpoint endpoint, CodeWriter codeWriter)
+    {
+        var serializerOptionsEmitted = false;
+        if (endpoint.Response?.IsSerializableJsonResponse(out var responseType) == true)
+        {
+            var typeName = responseType.ToDisplayString(EmitterConstants.DisplayFormat);
+            codeWriter.WriteLine("var serializerOptions = serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value.SerializerOptions ?? new JsonOptions().SerializerOptions;");
+            serializerOptionsEmitted = true;
+            codeWriter.WriteLine($"var responseJsonTypeInfo =  (JsonTypeInfo<{typeName}>)serializerOptions.GetTypeInfo(typeof({typeName}));");
+        }
+
+        foreach (var parameter in endpoint.Parameters)
+        {
+            ProcessParameter(parameter, codeWriter, ref serializerOptionsEmitted);
+            if (parameter is { Source: EndpointParameterSource.AsParameters, EndpointParameters: {} innerParameters })
+            {
+                foreach (var innerParameter in innerParameters)
+                {
+                    ProcessParameter(innerParameter, codeWriter, ref serializerOptionsEmitted);
+                }
+            }
+        }
+
+        static void ProcessParameter(EndpointParameter parameter, CodeWriter codeWriter, ref bool serializerOptionsEmitted)
+        {
+            if (parameter.Source != EndpointParameterSource.JsonBody && parameter.Source != EndpointParameterSource.JsonBodyOrService && parameter.Source != EndpointParameterSource.JsonBodyOrQuery)
+            {
+                return;
+            }
+            var typeName = parameter.Type.ToDisplayString(EmitterConstants.DisplayFormat);
+            if (!serializerOptionsEmitted)
+            {
+                serializerOptionsEmitted = true;
+                codeWriter.WriteLine("var serializerOptions = serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value.SerializerOptions ?? new JsonOptions().SerializerOptions;");
+            }
+            codeWriter.WriteLine($"var {parameter.SymbolName}_JsonTypeInfo =  (JsonTypeInfo<{typeName}>)serializerOptions.GetTypeInfo(typeof({parameter.Type.ToDisplayString(EmitterConstants.DisplayFormatWithoutNullability)}));");
```

Comment:
The nullability is only emitted from the type that we pass into the `typeof` call (`GetTypeInfo(typeof(Todo?))` vs `GetTypeInfo(typeof(Todo))`). The resolved type info still contains the correct annotation (`JsonTypeInfo<Todo?>`).

I didn't add any new tests since we already have coverage for body binding scenarios but [this test](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.JsonBody.cs#L79) covers reading from types with various annotations....